### PR TITLE
Adds KeyBoard Navigation to ObjectFilterDropDownFilterSelect ( #4365 )

### LINKED ISSUE #231: All selectable lists should be navigable using arrow keys.
# Current behavior

Some `selectable lists` cannot be navigated up or down using the keyboard:

### UX elements:

- View Menu
Ignore the 15+ views with identical names. 😅
CleanShot 2024-03-07 at 18 06 18

- Filter Menu
CleanShot 2024-03-07 at 18 06 00

- Sort Menu
CleanShot 2024-03-07 at 18 05 42

- Options Menu
CleanShot 2024-03-07 at 18 05 28

- Descending/Ascending menu:
image

- Relation Selection menu:
CleanShot 2024-03-07 at 18 07 53

# Desired behavior

We want all these menus to be navigable by arrow keys. Pressing "Enter" should select the focused item, and pressing "Escape" should close the menu or picker.

### LINKED ISSUE #4365: All selectable lists should be navigable using arrow keys.
# Current behavior

Some `selectable lists` cannot be navigated up or down using the keyboard:

### UX elements:

- View Menu
Ignore the 15+ views with identical names. 😅
CleanShot 2024-03-07 at 18 06 18

- Filter Menu
CleanShot 2024-03-07 at 18 06 00

- Sort Menu
CleanShot 2024-03-07 at 18 05 42

- Options Menu
CleanShot 2024-03-07 at 18 05 28

- Descending/Ascending menu:
image

- Relation Selection menu:
CleanShot 2024-03-07 at 18 07 53

# Desired behavior

We want all these menus to be navigable by arrow keys. Pressing "Enter" should select the focused item, and pressing "Escape" should close the menu or picker.

### PR DESCRIPTION
**Pull request by Faisal-imtiyaz123**
fixes twentyhq/core-team-issues#231

### CODE REVIEW

**Comment by Faisal-imtiyaz123** [comment]
@FelixMalfait To Reset the Selected Position upon closing filter dropdown , I was thinking of following a similar approach to ` SINGLE_ENTITY_SELECT_BASE_LIST` by using `useSelectableList` hook again in `DropDown.tsx`  and changing `handleResetSelectedPosition` to a apt name and then calling this inside the `useListenClickOutsideHook`. Is it fine?

**Comment by Faisal-imtiyaz123** [comment]
@FelixMalfait @lucasbordeau  Please review this commit. If it is fine I will implement navigation for other `dropdowns` as well

File: packages/twenty-front/src/modules/object-record/object-filter-dropdown/components/ObjectFilterDropdownFilterSelect.tsx [hunk 1]
```diff
@@ -53,7 +57,15 @@ export const ObjectFilterDropdownFilterSelect = () => {
   const { getIcon } = useIcons();
 
   const setHotkeyScope = useSetHotkeyScope();
-
+  const {isSelectedItemIdSelector} = useSelectableList(OBJECT_FILTER_DROPDOWN_ID)
+  const isSelectedItemId = (id:string)=> useRecoilValue(isSelectedItemIdSelector(id));
```
Comment (lucasbordeau) [inline packages/twenty-front/src/modules/object-record/object-filter-dropdown/components/ObjectFilterDropdownFilterSelect.tsx:61]: That's not a common pattern, prefer using a useRecoilCallback and extract it in an external file like useIsListSelectedItem

File: packages/twenty-front/src/modules/object-record/object-filter-dropdown/components/ObjectFilterDropdownFilterSelect.tsx [hunk 2]
```diff
@@ -53,7 +57,15 @@ export const ObjectFilterDropdownFilterSelect = () => {
...
+  const refactoredAvailableFilterDefinitions = [...availableFilterDefinitions]
```
Comment (lucasbordeau) [inline packages/twenty-front/src/modules/object-record/object-filter-dropdown/components/ObjectFilterDropdownFilterSelect.tsx:62]: refactoredAvailableFilterDefinitions => sortedSelectedFilterDefinitions ?

File: packages/twenty-front/src/modules/object-record/object-filter-dropdown/components/ObjectFilterDropdownFilterSelect.tsx [hunk 3]
```diff
@@ -64,18 +76,14 @@ export const ObjectFilterDropdownFilterSelect = () => {
...
-        {[...availableFilterDefinitions]
-          .sort((a, b) => a.label.localeCompare(b.label))
-          .filter((item) =>
-            item.label
-              .toLocaleLowerCase()
-              .includes(searchText.toLocaleLowerCase()),
-          )
+        {refactoredAvailableFilterDefinitions
           .map((availableFilterDefinition, index) => (
             <MenuItem
```
Comment (lucasbordeau) [inline packages/twenty-front/src/modules/object-record/object-filter-dropdown/components/ObjectFilterDropdownFilterSelect.tsx:83]: Can't you use `MenuItemSelect` instead ? Seems like it's a better match for this use case.

File: packages/twenty-front/src/modules/object-record/object-filter-dropdown/components/ObjectFilterDropdownFilterSelect.tsx (hunk 1 above)
Comment (Faisal-imtiyaz123) [reply packages/twenty-front/src/modules/object-record/object-filter-dropdown/components/ObjectFilterDropdownFilterSelect.tsx:61]: ok got it.

**Comment by Faisal-imtiyaz123** [comment]
@lucasbordeau Regarding the `3rd comment`, should i then use `MenuItemSelect` with `selected=false` in place of the current `MenuItem` . Also `MenuItemSelect` dosen't have `test-id` prop on it which is there on `MenuItem` currently.

**Comment by Faisal-imtiyaz123** [comment]
@lucasbordeau Regarding `comment 3` should I then use `MenuItemSelect` with `selected=false` instead of the current `MenuItem`. Also `MenuItemSelect` dosen't have `testId` which is there on `MenuItem`. So if it is not important then removing it should be fine.

**Comment by Faisal-imtiyaz123** [comment]
@lucasbordeau Regarding `comment 1` what should be the best place to put the file `useIsListSelectedItem.ts` in?

**Comment by lucasbordeau** [comment]
@Faisal-imtiyaz123 
- Use MenuItemSelect => yes
- File location, in modules/ui/layout/selectable-list/hooks

**Comment by Faisal-imtiyaz123** [comment]
@lucasbordeau Regarding `Comment 1`, I don't think using a recoil callback will help with any performance since the number of renders wont't change also the `callback` should run if `id` changes so I will require `selectedItemIdState` , i don't think it is possible write a callback using `isSelectedItemIdSelector` because then we don't have access to `id`

**Comment by lucasbordeau** [comment]
@Faisal-imtiyaz123 Yes you can extract without using a recoil callback but the main point is to extract this in an external hook at least.

**Comment by Faisal-imtiyaz123** [comment]
@lucasbordeau You may review it.

**Comment by Faisal-imtiyaz123** [comment]
@lucasbordeau I earlier wrote the same code using `useRecoilCallback` but it won't work because the callBack should have `listItemId` in dependancy . You wouldn't be able to navigate now .

**Comment by lucasbordeau** [comment]
@Faisal-imtiyaz123 The only dependency is selectedListId, listItemId is a param of the recoil callback, this what you need right ?

**Comment by Faisal-imtiyaz123** [comment]
@lucasbordeau Please once try navigating the `Filter Menu` . You won't be able to.

**Comment by lucasbordeau** [comment]
We'll see if we can factorize even more the isSelected logic, maybe in a component SelectableListMenuItem ?

**Comment by lucasbordeau** [comment]
@Faisal-imtiyaz123 Yes you were right, I tried a different approach myself to see what works best.

**Comment by lucasbordeau** [comment]
Could you take care of resetting the position when we close the dropdown ? Otherwise the position is at the same place.

It would also be interesting to use a useScopedHotkeys on enter to validate the selected item.

**Comment by Faisal-imtiyaz123** [comment]
@lucasbordeau In `SingleEntitySelectMenuItems` where keyboard navigation is enabled, the `selected listItem` dosen't `reset` when dropdown is closed. I think it should be a default feature with every `SelectableList`  to  reset `selected Item` when closed. So maybe we should handle this in the `SelectableList code` itself instead of trying for `individual components` and there is also a bug that if you click on `filter sort or option`, then the `pencil icon` to trigger `SingleEntitySelectMenuItems` disappears. I plan to work on that also.

**Comment by lucasbordeau** [comment]
@Faisal-imtiyaz123 Yes I agree, whatever can be factorized in the selectable list module should be. Please keep in mind that you can create wrapping components like I did here, and if you need to create one for each list item type that's ok, it's better to have a lot of small components that do a very specific thing even just wrapping a UI component into selectable list logic.

**Comment by Faisal-imtiyaz123** [comment]
@lucasbordeau Sorry but I didn't understand what you meant by this ```It would also be interesting to use a useScopedHotkeys on enter to validate the selected item.```. To be exact I didn't get this ``` validate the selected item```.

**Comment by Faisal-imtiyaz123** [comment]
@lucasbordeau You may review it.

**Comment by lucasbordeau** [comment]
I've made some minor changes and also added the <SelectableItem> that is required when the list is overflowing.

Also I refactored resetSelectedItem and used it when we close instead of when we open because it creates a glitch on open. 

I also removed selectable list module calls from the internal dropdown component which should be separated from the selectable list module. 

You can proceed with the next dropdown ! Thanks !

**Review by lucasbordeau — APPROVED** [review]

**Pull request merged**